### PR TITLE
Checkout: Test using Google Places API for smoother checkout

### PR DESCRIPTION
**Pull request by klimeryk**
#### Changes proposed in this Pull Request

We are required to collect address contact info when registering a domain - however, the data has to be in the correct format, it's a PITA to fill out, etc. Using the Google Places API could help with that and ensure consistent and valid data.

The way I've approached it was to hide the address fields at first and require the user to use the Google Places API to search for their address first. If they can't find it, only then they can choose to fill it out manually.

A/B test name: `placesApiInCheckout`

#### Testing instructions

Try adding a domain to your site (signup or existing one, either way is the same). During checkout, you'll be prompted to enter your contact details.

Make sure that for the `original` variation there are no regressions (especially for new users, when the contact form is empty).

Try different addresses you could think of - especially in exotic countries, etc. (Google Places API has a slightly weird way of returning addresses and it might need tweaking)

The last suggestion is always a placeholder one that allows you fill out the form manually. To help figure out if/how often users select it, I've added a `calypso_contact_information_manual_address_click` event.

#### Screenshots

Screenshot 2019-08-21 at 20 48 59
Screenshot 2019-08-26 at 15 53 27

### CODE REVIEW

**Comment by klimeryk** [comment]
Thanks for the feedback, @fditrapani! I wasn't happy with the "bigger than others" search field for the address, but that was because it was using the `Search` component (hence the 🔍 icon, etc.). But your suggestion totally makes sense - I've now made it a normal input, just like the other ones in the form. And made it a more integral part of it.

As for the second line of address - it's a legacy thing and I think we have it there because it might be useful for some addresses (and seems most/many forms online I've seen have it)... but I honestly don't see the need for it (and it makes the implementation easier), so I've removed it for now 🤷‍♂ Let's see if we get any feedback from customers during the test about it.

Pushed the new changes - let me know how they look now 🙇

**Comment by fditrapani** [comment]
Looks much better @klimeryk! For some reason, I still can't see any suggestions showing on my end. Do I need to sandbox or anything to see it work?

This is what I see in the console:
image

**Comment by klimeryk** [comment]
@fditrapani Sorry for a slow follow-up. Looks like the root issue you're facing is that the Google API key is not whitelisted for *.calypso.live - and that's causing Google to reject it. Could you try running the code locally and access it via `calypso.localhost:3000` as you'd normally do during development? Feel free to ping me on Slack if you encounter further issues bootstrapping the project 🙇

**Comment by klimeryk** [comment]
I've updated the PR to show the correct attribution, as [per Google's policy](https://developers.google.com/places/web-service/policies#powered). It's a bit bare-boned, so appreciate any feedback on it:

Screenshot 2019-08-26 at 15 53 27

Wanted to mimic the style that Google uses:

Screenshot 2019-08-26 at 19 48 13

(live example here: https://developers.google.com/maps/documentation/javascript/examples/places-autocomplete)

But seems our `Card` components use `box-shadow` for some reason and did not see an easy way of specifying no border on the bottom of one of them without switching to borders... which changes their sizes and since we use `Card`s all around Calypso, that's a change I'd rather not risk 😅 But maybe I'm missing some CSS wizardry that can be done.

**Comment by klimeryk** [comment]
Thank you for another stab at this and detailed feedback, @fditrapani!

> When I select an address, it swaps the address number and street name around ie: 906 Sandy Forest Place becomes Sandy Forest Place 906

Good catch, but unfortunately I don't have an easy fix for this. The problem is that the initial so called "prediction" (the suggestion during the autocomplete phase) is sent from Google in a merged format like this:

```
   "predictions" : [
      {
         "description" : "906 Sandy Forest Place, Gloucester, ON, Canada",
         ....
       }
     ]
```

However, that's only useful for a quick display/preview of the address. Upon clicking on the suggestion, we fetch the full place details via its ID and get much more... but that's also the downside:

```
"address_components" : [
         {
            "long_name" : "906",
            "short_name" : "906",
            "types" : [ "street_number" ]
         },
         {
            "long_name" : "Sandy Forest Place",
            "short_name" : "Sandy Forest Pl",
            "types" : [ "route" ]
         },
         {
            "long_name" : "Gloucester",
            "short_name" : "Gloucester",
            "types" : [ "sublocality_level_1", "sublocality", "political" ]
         },
         {
            "long_name" : "Ottawa",
            "short_name" : "Ottawa",
            "types" : [ "locality", "political" ]
         },
         {
            "long_name" : "Ottawa Division",
            "short_name" : "Ottawa Division",
            "types" : [ "administrative_area_level_2", "political" ]
         },
         {
            "long_name" : "Ontario",
            "short_name" : "ON",
            "types" : [ "administrative_area_level_1", "political" ]
         },
         {
            "long_name" : "Canada",
            "short_name" : "CA",
            "types" : [ "country", "political" ]
         },
         {
            "long_name" : "K1V 1R5",
            "short_name" : "K1V 1R5",
            "types" : [ "postal_code" ]
         }
      ]
```

Now, everything is nicely separated... but there's no indication how these "pieces" should go together.

We _can_ request the formatted address from the API as well ([`formatted_address` field](https://developers.google.com/places/web-service/details#fields)), but it includes the city, state and country as well: `906 Sandy Forest Place, Gloucester, ON, Canada` and there is even a warning in the docs:

> Do not parse the formatted address programmatically. Instead you should use the individual address components, which the API response includes in addition to the formatted address field.

And I'd heed to it, because https://www.mjt.me.uk/posts/falsehoods-programmers-believe-about-addresses/ (one of my all time favourite websites/articles).

We could add some logic that for these countries the street number goes first, while for others it goes last, but not sure if it's worth it (and with our scale, not sure if it's possible). We definitely should be on the lookout for any issues reported by users/HEs around this during the test, of course. Worst case scenario, we can do the fix for US (and Canada? see, I'm not even sure what the rules are 😅), so that we eliminate the biggest point of confusion.

> As soon as I start typing the Country field appears. It can probably stay hidden until all the other fields are displayed.

Aye, noticed that too, but was hoping it'd not be an issue 😅 It's because it's hard to tell in which state are we - landed on the checkout with saved data or have some fresh input, but no selected location. But thinking about this more, I figured out a solution to it without complicating the code too much. Thanks for the push!

> If I want to try navigate with my keyboard, when I press tab inside the address field it takes me to the country field instead of the search results.

Good catch, totally missed that. Made some changes in 55253573c7012273f8a55cc56ea54180c2541eac that should address it.

> If I try enter the address manually, I see the address field and the country field. It's not clear to me that the other fields will display after I pick a country. We could hide the address field and update the country field placeholder text to say "Start by selecting your country".

Hmmm, good point. It's the result of how this form requires the country to be selected first (so that we know in which format to show the postal code and city fields - and the optional state field), but in this new approach it makes less sense.

Pushed a change that should hopefully fix this (as well as made the validation a bit more sane and less trigger-happy).

File: client/components/domains/contact-details-form-fields/style.scss [hunk 1]
```diff
@@ -93,7 +93,24 @@
...
+.contact-details-form-fields__field.location-search,
 .form__hidden-input .form__hidden-input {
 	margin-top: 0;
 }
 
+.contact-details-form-fields__field.location-search {
+	.address-1 {
+		position: relative;
+
+		&:focus {
```
Comment (fditrapani) [inline client/components/domains/contact-details-form-fields/style.scss]: this doesn't seem to be getting picked up. We could leave it on all the time. Just looking into why it's not working or other options.

**Comment by klimeryk** [comment]
Thank you for all your help on this, @fditrapani! 🙌 Really happy with all your suggestions and improvements 🙇 

> How are we constructing the address that shows in the field now. Is that something coming from Google?

I just build it based on what I think is the street name and then the house/apartment number. See this line: https://github.com/Automattic/wp-calypso/pull/34940/files#diff-9868062561d61e0868549c5cc9e305a1R375

Basically, I search for `street_address`, `route` and `street_number` in the response from Google and then join them together (those that are present). This way it's more flexible (depending on the search query, some parts might be missing), but the order in which I join them is always the same.

**Comment by fditrapani** [comment]
Thank you for making this work happen @klimeryk!  

> I just build it based on what I think is the street name and then the house/apartment number. 

Cool, thanks for the background!

**Pull request merged**